### ISSUE TITLE
Memoized encoding does not properly handle repeated field operations

### ISSUE THREAD

**Issue by goddardc** [ruby-protobuf/protobuf#304]
The changes introduced in https://github.com/ruby-protobuf/protobuf/pull/293, performing operations on repeated fields is no longer handled properly after the memoized encoding has been computed.

```ruby
a.encode
a.some_repeated_field << "new-value"
a.encode # doesn't contain "new-value"
```

**Comment by liveh2o** [comment]
Dug into this and it may end up being a deal breaker for the memoized encoding change. Because the field array we are pushing into doesn't have any concept of a message, let alone the specific message instance the field belongs to, there is no way to clear the memoized encoded message. It works for other field types because the memoized value is being cleared in the setter before we call into the field. The field is not involved at all. I'm thinking we may need to revert the memoization change until we can find a better solution.

// @abrandoned @film42 @mmmries

**Comment by liveh2o** [comment]
Released in [v3.6.7](https://github.com/ruby-protobuf/protobuf/releases/tag/v3.6.7).
